leetcode/medium/15._3Sum.py

```diff
@@ -0,0 +1,74 @@
+'''
+Link to problem: https://leetcode.com/problems/3sum/
+Loom explaining solution you wrote (5m):
+Solution:
+Description
+Key ideas:
+ - Be plain in terms of approach, with unfamiliar question write out the most 
+   intuitive brute-force close solution first.
+ - Optimize the problem with Big-O
+Difficulty: Medium (was hard), mainly due to optimizing timeout.
+'''
+
+def containsInSorted(arr: List[int], target: int) -> bool:
+    lo, high = 0, len(arr)
+    while lo < high:
+        mid = (lo + high) // 2
+        if arr[mid] == target:
+            return True
+        elif arr[mid] > target:
+            high = mid
+        else:
+            lo = mid + 1
+    return 0 <= lo < len(arr) and arr[lo] == target
+
+'''
+Double for-loop + binary search, O(N^2 log N)
+De-dup through set
+'''
+class BinarySearch:
+    def threeSum(self, nums: List[int]) -> List[List[int]]:
+        nums = sorted(nums)
+
+        res = set()
+        for first_idx, first_val in enumerate(nums):
+            if first_val > 0:
+                break
+            if first_idx > 0 and nums[first_idx - 1] == first_val:
+                continue
+
+            for second_idx, second_val in enumerate(nums[first_idx + 1:], first_idx + 1):
```

Comment:
You need to be careful when slicing lists as it adds additional time complexity and memory. You should be able to solve this problem without slicing. here's how. 

1. Sort the list 
2. Iterate index `i` throughout the entire nums. 
3. Perform binary search on indices `i+1:` You can pass the entire `nums` into your binary search function, but you can pass the starting index as well. This avoids the need to slice the array.